### PR TITLE
Journaling bloom filter crate in util

### PR DESCRIPTION
**Pull request by NikVolf**

### CODE REVIEW

File: util/bloom/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+/// Bloom journal
+/// Returns the tuple of (bloom part index, bloom part value) where each one is representing
+/// an index of bloom parts that was updated since the last drain
+pub struct BloomJournal {
+    pub hash_functions: u32,
+    pub entries: Vec<(usize, u64)>,
+}
+
```
Comment (rphmeier) [inline util/bloom/src/lib.rs]: tests could be in a separate `tests` module

**Comment by rphmeier** [comment]
Needs the license from the original work: https://github.com/jedisct1/rust-bloom-filter/blob/master/LICENSE

(is this compatible with our GPL? @tjsaw)

**Comment by NikVolf** [comment]
@rphmeier 
don't think any license talks valid
it is 3 or 4 lines left from the original crate

File: util/bloom/src/lib.rs (hunk 1 above)
Comment (NikVolf) [inline util/bloom/src/lib.rs]: done, though it's not much of a convention

**Comment by arkpar** [comment]
This is BSD-2-Clause. Pretty sure it is compatible

File: util/bloom/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+	}
+
+	pub fn get(&self, index: usize) -> bool {
+		let e_index = index / 64;
+		let bit_index = index % 64;
+		self.elems[e_index] & (1 << bit_index) != 0
+	}
+
+	pub fn drain(&mut self) -> Vec<(usize, u64)> {
+		let journal = self.journal.drain().collect::<Vec<usize>>();
```
Comment (arkpar) [inline util/bloom/src/lib.rs]: No need for a temporary vector here

File: util/bloom/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+		journal.iter().map(|idx| (*idx, self.elems[*idx])).collect::<Vec<(usize, u64)>>()
+	}
+
+	pub fn how_full(&self) -> f64 {
```
Comment (arkpar) [inline util/bloom/src/lib.rs]: Should be called "saturation" or something like that

File: util/bloom/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+			bitmap: bitmap,
+			bitmap_bits: bitmap_bits,
+			k_num: k_num,
+			sips: sips,
+		}
+	}
+
+	/// Initializes bloom filter from saved state
+	pub fn from_parts(parts: &[u64], k_num: u32) -> Bloom {
+		let bitmap_size = parts.len()*8;
```
Comment (arkpar) [inline util/bloom/src/lib.rs]: spacing around '*'

File: util/bloom/src/lib.rs (hunk 2 above)
Comment (NikVolf) [inline util/bloom/src/lib.rs]: not a simple change since drain holds `&mut` ref to `self`
but i will try something

Comment (arkpar) [inline util/bloom/src/lib.rs]: `let journal = mem::replace(&self.journal, HashMap::new()).into_iter()`

Comment (rphmeier) [inline util/bloom/src/lib.rs]: or explicitly borrow the other field of self outside of the closure so the closure doesn't capture `self`.

Comment (NikVolf) [inline util/bloom/src/lib.rs]: done

File: util/bloom/src/lib.rs (hunk 3 above)
Comment (NikVolf) [inline util/bloom/src/lib.rs]: done

File: util/bloom/src/lib.rs (hunk 4 above)
Comment (NikVolf) [inline util/bloom/src/lib.rs]: done

**Pull request merged**